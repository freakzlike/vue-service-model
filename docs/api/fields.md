# Fields API

```typescript
class Field {
  // Constructor takes field definition
  constructor (def: FieldDef, fieldBind?: FieldBind)

  // Clone field instance
  public clone (): Field

  // Returns field name (Which has been set as key at fieldsDef.
  // Will throw FieldNotBoundException in case field has not been bound
  public get name (): string

  // Returns either attributeName from FieldDef or field name
  public get attributeName (): string

  // Returns field definition
  public get definition (): FieldDef

  // Assigned model
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get model (): BaseModel

  // Return bound data or data from bound model
  // Will throw FieldNotBoundException if field is not bound to data or model 
  public get data (): Dictionary<any>

  // Property mapper for getValue()
  public get value (): any

  // Returns async field value from data by calling valueGetter with data of assigned model
  public async getValue (): Promise<any>

  // Property mapper for setValue()
  public set value (value: any)

  // Sets field value to model data by calling valueSetter with data of assigned model
  public setValue (value: any): void

  // Parses raw value with valueParser and sets field value by calling valueSetter
  public async setParseValue (rawValue: any): Promise<any>

  // Returns async field label from field definition
  public get label (): Promise<string>

  // Returns async field options with validation and default values depending on field type
  public get options (): Promise<FieldTypeOptions>

  // Validate field options and set default values depending on field type
  protected async validateOptions (options: FieldTypeOptions): Promise<FieldTypeOptions>

  // Returns boolean whether field is a primary key
  public get isPrimaryKey (): boolean

  // Returns boolean whether attribute is an nested data structure
  public get isNestedAttribute (): boolean

  // Retrieve value from data structure according to attributeName
  // Uses nested syntax from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  // Will return null if value is not available
  public valueGetter (data: any): any

  // Set value to data by using attributeName
  // Will create nested structure from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  public valueSetter (value: any, data: Dictionary<any>): void

  // Parse a raw value and return the parsed value with valid data type
  public async valueParser (rawValue: any): Promise<any>

  // Map value from a data structure to another data structure.
  // Uses valueGetter and valueSetter
  public mapFieldValue (fromData: Dictionary<any>, toData: Dictionary<any>): void
}
```

## RenderableField

```typescript
class RenderableField extends Field {
  // Display component to render when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public get displayComponent (): Promise<ComponentModule>

  // Async function to prepare before displayRender gets called
  // Can return any data which needs to be resolved for displayRender
  // renderProps can be passed from DisplayField component 
  public async prepareDisplayRender (renderProps?: object | null): Promise<any>

  // Simple Vue render function when using default displayComponent when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public displayRender (h: CreateElement, renderData: any): VNode

  // Input component to render when showing input for field with <input-field/>
  // For more information see Field - Rendering 
  public get inputComponent (): Promise<ComponentModule>

  // Async function to prepare before inputRender gets called
  // Can return any data which needs to be resolved for inputRender
  // renderProps can be passed from InputField component
  public async prepareInputRender (inputProps: InputProps, renderProps?: object | null): Promise<InputRenderData>

  // Simple Vue render function when using default inputComponent for input of field value with <input-field/>
  // For more information see Field - Rendering 
  public inputRender (h: CreateElement, renderData: InputRenderData): VNode
}
```

## FieldDef

```typescript
interface FieldDef {
  // String which key should be used to retrieve value from.
  // See Attribute name for more information
  // Optional: default uses key from fieldsDef
  attributeName?: string

  // Label of field. See Field label for more information
  // Optional: Can either be a string, function or promise
  label?: LazyValue<string>

  // Boolean flag whether field is a primary key
  // Optional: default is false
  primaryKey?: boolean

  // Optional field type specific options.
  // Can either be an object, a function or a promise.
  // See field Types
  options?: LazyValue<FieldTypeOptions>
}
```

## FieldBind

```typescript
interface FieldBind {
  // Field name which has been set as key at fieldsDef
  name?: string

  // Model to bind field to
  // Beware of manual bind your field to a model afterwards.
  // The field will not be available by model.fields or model.getField()
  model?: BaseModel

  // Data to retrieve field value from
  // If not set then model.data will be used
  data?: Dictionary<any>

  // Value which should be bound to field
  // When using value then model and data will be ignored
  // A custom data object will be created to keep the field reactive
  value?: any
}
```

## InputProps

```typescript
interface InputProps {
  // Input flag for disabled input
  disabled: boolean

  // Input flag for readonly input
  readonly: boolean
}
```

## InputRenderData

```typescript
interface InputRenderData {
  // Resolved value of field
  value: any

  // Input properties
  inputProps: InputProps
}
```