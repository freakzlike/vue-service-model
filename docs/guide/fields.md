# Fields

[[toc]]

A `Field` provides information and methods to work with your data. An information can be the label which is used to
display a name of the field for the user (e.g. as column header). A [`RenderableField`](/guide/fields.html#field-vs-renderablefield)
inherits from `Field` and additionally provides a handling to render to field value to the user or render an input field.

## Field definition

When instantiating a new field you can provide a specific definition for the field.

```js
class Album extends BaseModel {
  [...]

  static fieldsDef = {
    id: new Field({primaryKey: true}),
    title: new Field()
  }
}
```

Field definition structure:
```js
{
  // String which key should be used to retrieve value from.
  // See Attribute name for more information
  // Optional: default uses key from fieldsDef 
  attributeName: 'title',

  // Label of field. See Field label for more information
  // Optional: Can either be a string, function or promise
  label: () => Promise.resolve('Title'),

  // Boolean flag whether field is a primary key
  // Optional: default is false
  primaryKey: true,

  // Optional field type specific options.
  // Can either be an object, a function or a promise.
  // See field Types
  options: {}
}
```

### Attribute name (`attributeName`)

By default the key of your field in your `fieldsDef` (e.g. `first_name`) will be used to retrieve the value from the model data.
You can also set the `attributeName` when instantiating the field. It is also possible to access nested data when using dot-notation in `attributeName`.
If you need a more specific way to retrieve the value of a field from your data then have a look at [Custom/Computed fields](#customcomputed-fields).

```js
class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    name: new Field({attributeName: 'username'}),
    address_city: new Field({attributeName: 'address.city'}),
    address_street: new Field({attributeName: 'address.street.name'})
  }
}

const myObj = new MyModel({
  username: 'joe_bloggs',
  address: {
    city: 'New York',
    street: {
      name: 'Fifth Avenue'
    } 
  }
})

await myObj.val.name // output: joe_bloggs
await myObj.val.address_city // output: New York
await myObj.val.address_street // output: Fifth Avenue
```

### Field label (`label`)

With the `label` property you can set a descriptive name for your field. The `label` can either be a `string` or a `function` which should return a `string` or a `Promise`.
You can access your label with the `label` property of your field instance which will always return a `Promise`.

```js
class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    first_name: new Field({
      label: 'First name'
    })
  }
}

[...]

const firstNameField = myObj.getField('first_name')

await firstNameField.label // output: First name
```

## Standalone Field

In some cases you just want to work with a single field and don't need a model.
So it is possible to create a standalone field by directly passing an initial value.

```js
// Pass initial value when creating a new instance
const myField = new Field({label: 'Description'}, {value: 'My description'})

await myField.value // output: My description
```

::: tip note
The value will be stored in a data object to keep the field reactive. Your can retrieve the data object with `myField.data`
:::

You can also directly pass a data object instead of an initial value

```js
const data = {description: 'My description'}
// Bind data when creating a new instance
const myField = new Field({attributeName: 'description'}, {data: data})

await myField.value // output: My description
```

You can also use the [components](/guide/components.html) to render your standalone field. Be sure to use a [`RenderableField`](/guide/fields.html#field-vs-renderablefield).

```vue
<template>
  [...]
    <display-field :field="myField"/>
  [...]
</template>
```

## Custom/Computed fields

In case you want to define your own field class you just need to extend from `Field`. By overwriting the `valueGetter` method you are able to map the field value by yourself and create computed values.

```js
class FullNameField extends Field {
  valueGetter (data) {
    return data ? `${data.first_name} ${data.last_name}` : null
  }
}

class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    full_name: new FullNameField()
  }
}

const myObj = new MyModel({
  first_name: 'Joe',
  last_name: 'Bloggs'
})

await myObj.val.full_name // output: Joe Bloggs
```

## Value parsing

To avoid unexpected data types of your field value, then you should use the `setParseValue` method of your field. This will
ensure that your Field (e.g. `IntegerField`) always contains a `number`.

```js
// Set value without parsing
myIntegerField.value = '5'
// -> Field value will be string with value '5'

// Parse and set value
myIntegerField.setParseValue('5')
// -> Field value will be number with value 5
```

You can customize your parsing logic by overwriting the `valueParser`.

```js
class StringField extends Field {
  valueparser (rawValue) {
    return String(rawValue)
  }
}
```

## `Field` vs `RenderableField`

In case you don't need the rendering functions, then you can use the `Field` class and avoid unnecessary code which will make your bundle smaller.
The `RenderableField` provides additional methods like `displayRender` to allow rendering of the field.

## Rendering

::: warning
Be sure to use a field which inherits from [`RenderableField`](/guide/fields.html#field-vs-renderablefield). The default `Field` class does not support rendering.
:::

By using the [`DisplayField`](/guide/components.html#displayfield) component you can render the value of a field for displaying purpose
and [`InputField`](/guide/components.html#inputfield) when you want to render the input component for this field.
To customize the different output which should be rendered you can either set a `displayRender`/`inputRender` or a custom `displayComponent`/`inputComponent`.

The `displayRender`/`inputRender` can be used for small changes of the output and is a simple render function (See [Vue.js render function](https://vuejs.org/v2/guide/render-function.html)).

```js
class RedTextField extends RenderableField {
  // Render a red text for display
  displayRender (h, resolvedValue) {
    return h('span', {
      style: {
        color: 'red'
      }
    }, resolvedValue)
  }

  // Render an text input field for input
  inputRender (h, { value, inputProps }) {
    // Common input properties
    const { disabled, readonly } = inputProps

    return h('input', {
      attrs: {
        type: 'text',
        // Set current value to input
        value,
        // Implement common input properties
        disabled,
        readonly
      },
      on: {
        input: (event) => {
          // Parse and set input value to field.value
          this.setParseValue(event.target.value)
        }
      }
    })
  }
}
```

::: tip
In case you need to achieve more reactivity, then take a look at [Integration of `vue-async-computed`](/guide/installation.html#integration-of-vue-async-computed).
:::

### Resolving asynchronous values before rendering

In some cases you need to prepare or fetch other data before the field should be rendered. This can be done with the methods `prepareDisplayRender`/`prepareInputRender`.
They will be called before `displayRender`/`inputRender` and will give any async data to as 2. argument to them.

```js
class AlbumTitleField extends RenderableField {
  // Do any asynchronous operations before rendering with displayRender
  async prepareDisplayRender (renderProps) {
    const albumId = await this.value
    const album = await Album.objects.detail(albumId)
    return album.val.title
  }

  // Render the asynchronous fetched data
  // renderData will contain the value returned by prepareDisplayRender
  displayRender (h, renderData) {
    return h('span', renderData)
  }
}
```

::: tip
You can optionally use the `renderProps` argument which can be used to pass additional properties from [`DisplayField`](/guide/components.html#displayfield) component to [`prepareDisplayRender`](/guide/components.html#passing-additional-properties-to-preparedisplayrender)
:::

### Custom field component

In case you need to do more specific rendering you can also set your own component which will be rendered when using [`DisplayField`](/guide/components.html#displayfield) on your custom field.
If you want to change the input component you can extend from `BaseInputFieldRender` and overwrite the `inputComponent` method of your field.

```js
// CustomField.js
class CustomField extends RenderableField {
  get displayComponent () {
    return import('./CustomFieldComponent')
  }
}

// CustomFieldComponent.vue
<template>
  <span>{{ fieldValue }}</span>
</template>

<script>
  import {BaseDisplayFieldRender} from 'vue-service-model'

  export default {
    extends: BaseDisplayFieldRender,
    data: () => ({
      fieldValue: null    
    }),
    created () {
      this.field.value.then(value => (this.fieldValue = value))      
    }
  }
</script>
```

To keep it simple and reactive you can install the package [`vue-async-computed`](https://github.com/foxbenjaminfox/vue-async-computed).
See [Integration of `vue-async-computed`](/guide/installation.html#integration-of-vue-async-computed).

```js
[...]

<script>
  import {BaseDisplayFieldRender} from 'vue-service-model'

  export default {
    extends: BaseDisplayFieldRender,
    asyncComputed: {
      fieldValue () {
        return this.field.value      
      }  
    }
  }
</script>
```
