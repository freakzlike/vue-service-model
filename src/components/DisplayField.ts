import { VNode, DefineComponent, defineComponent, h, toRaw } from 'vue'
import FieldPropsMixin from './FieldPropsMixin'
import LoadingSlotMixin from './LoadingSlotMixin'
import { configHandler } from '../utils/ConfigHandler'

export interface ComponentData {
  displayComponent: DefineComponent | null
}

/**
 * Main component to display value of a field
 */
export default defineComponent({
  name: 'DisplayField',
  inheritAttrs: false,
  mixins: [FieldPropsMixin, LoadingSlotMixin],

  data: (): ComponentData => ({
    displayComponent: null
  }),

  // TODO: vue-async-computed
  // asyncComputed: {
  //   async displayComponent (): Promise<DefineComponent | null> {
  //     configHandler.checkWarningUseAsyncComputed()
  //     return this.resolveDisplayComponent()
  //   }
  // },

  watch: {
    fieldObj () {
      this.setResolveDisplayComponent()
    }
  },

  created () {
    this.setResolveDisplayComponent()
  },

  methods: {
    async setResolveDisplayComponent () {
      if (!configHandler.useAsyncComputed()) {
        this.displayComponent = await this.resolveDisplayComponent()
      }
    },

    async resolveDisplayComponent () {
      if (this.fieldObj) {
        const componentModule = await this.fieldObj.displayComponent
        return componentModule.default
      } else {
        return null
      }
    }
  },

  render (): VNode {
    if (this.displayComponent && this.fieldObj) {
      return h(toRaw(this.displayComponent), {
        field: this.fieldObj,
        renderProps: this.renderProps
      })
    } else {
      return this.renderLoading()
    }
  }
})
