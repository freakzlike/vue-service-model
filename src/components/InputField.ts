import { VNode, defineComponent, DefineComponent, h, toRaw } from 'vue'
import FieldPropsMixin from './FieldPropsMixin'
import LoadingSlotMixin from './LoadingSlotMixin'

interface ComponentData {
  inputComponent: DefineComponent | null
}

/**
 * Main component to display value of a field
 */
export default defineComponent({
  name: 'InputField',
  inheritAttrs: false,
  mixins: [FieldPropsMixin, LoadingSlotMixin],

  props: {
    disabled: {
      type: Boolean,
      default: false
    },
    readonly: {
      type: Boolean,
      default: false
    }
  },

  data: (): ComponentData => ({
    inputComponent: null
  }),

  // TODO: vue-async-computed
  // asyncComputed: {
  //   async inputComponent (): Promise<DefineComponent | null> {
  //     configHandler.checkWarningUseAsyncComputed()
  //     return this.resolveInputComponent()
  //   }
  // },

  watch: {
    fieldObj () {
      this.setResolveInputComponent()
    }
  },

  created () {
    this.setResolveInputComponent()
  },

  methods: {
    async setResolveInputComponent () {
      // if (!configHandler.useAsyncComputed()) {
      this.inputComponent = await this.resolveInputComponent()
      // }
    },

    async resolveInputComponent () {
      if (this.fieldObj) {
        const componentModule = await this.fieldObj.inputComponent
        return componentModule.default
      } else {
        return null
      }
    }
  },

  render (): VNode {
    if (this.inputComponent && this.fieldObj) {
      return h(toRaw(this.inputComponent), {
        field: this.fieldObj,
        renderProps: this.renderProps,
        disabled: this.disabled,
        readonly: this.readonly
      })
    } else {
      return this.renderLoading()
    }
  }
})
