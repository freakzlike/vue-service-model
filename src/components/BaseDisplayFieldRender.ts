import { defineComponent, reactive, toRefs, watch, VNode } from 'vue'
import props, { DisplayComponentProps } from '../mixins/DisplayComponentPropsMixin'
import { Field } from '@/fields/Field'

export type RenderField = (field: Field, renderData: any) => VNode
export type ResolveRenderData = (field: Field, renderProps: object | null) => Promise<any>

export const BaseDisplayFieldRender = (
  props: DisplayComponentProps,
  renderField: RenderField,
  resolveRenderData: ResolveRenderData
) => {
  const { field, renderProps } = toRefs(props)

  const renderData: { data: any, resolved: boolean } = reactive({ data: null, resolved: false})

  const setResolveRenderData = async () => {
    renderData.data = await resolveRenderData(field.value, renderProps.value)
    renderData.resolved = true
  }

  const renderIfResolved = () => renderData.resolved ? renderField(field.value, renderData.data) : undefined

  setResolveRenderData()

  watch(field, setResolveRenderData)
  watch(field.value.data, setResolveRenderData, {deep: true})
  watch(renderProps, setResolveRenderData)

  return {
    field,
    renderProps,
    renderData,
    resolveRenderData,
    setResolveRenderData,
    renderField,
    renderIfResolved
  }
}

export default defineComponent({
  name: 'BaseDisplayFieldRender',
  inheritAttrs: false,
  props,

  setup (props) {
    const renderField: RenderField = (field, renderData) => field.displayRender(renderData)
    const resolveRenderData: ResolveRenderData = (field, renderProps) =>
      field.prepareDisplayRender(renderProps)

    const { renderIfResolved } = BaseDisplayFieldRender(<DisplayComponentProps> props, renderField,
      resolveRenderData)
    return renderIfResolved
  },
})
