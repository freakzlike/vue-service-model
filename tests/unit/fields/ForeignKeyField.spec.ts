import Vue, { CreateElement } from 'vue'
import { mount } from '@vue/test-utils'
import { BaseModel } from '@/models/BaseModel'
import { ServiceModel } from '@/models/ServiceModel'
import { Field } from '@/fields/Field'
import { CharField } from '@/fields/CharField'
import { ForeignKeyField, ForeignKeyFieldOptions } from '@/fields/ForeignKeyField'
import { InvalidFieldOptionsException, RequiredFieldOptionsException } from '@/exceptions/FieldExceptions'
import DisplayField from '@/components/DisplayField'
import { waitRender } from '../../testUtils'
import { ComponentModule } from '@/types/components'
import BaseDisplayFieldRender from '@/components/BaseDisplayFieldRender'
import InputField from '@/components/InputField'

describe('fields/ForeignKeyField', () => {
  class User extends ServiceModel {
    static fieldsDef = {
      id: new Field({ primaryKey: true }),
      name: new (class extends CharField {
        // Custom displayComponent to test DisplayField usage
        get displayComponent (): Promise<ComponentModule> {
          return Promise.resolve({
            default: Vue.extend({
              props: {
                field: {
                  type: Object,
                  required: true
                }
              },
              render (h: CreateElement) {
                return h('div', this.field.data.name)
              }
            })
          })
        }
      })()
    }
  }

  const userList = [
    new User({ id: 1, name: 'User 1' }),
    new User({ id: 2, name: 'User 2' }),
    new User({ id: 3, name: 'User 3' })
  ]
  const user = userList[0]

  class TestModel extends BaseModel {
    static fieldsDef = {
      user: new ForeignKeyField({
        options: {
          model: User,
          fieldName: 'name'
        }
      })
    }
  }

  const model = new TestModel({ user: user.pk })
  const field = model.getField('user') as ForeignKeyField

  describe('get value', () => {
    it('should fetch foreign instance', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => user)

      const value = await model.val.user

      expect(value).toBe(user)

      expect(mockObjectsDetail).toBeCalledTimes(1)
      expect(mockObjectsDetail.mock.calls[0]).toEqual([user.pk])
      mockObjectsDetail.mockRestore()
    })

    it('should not fetch on null value', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation()

      const model = new TestModel()
      const value = await model.val.user

      expect(value).toBeNull()

      expect(mockObjectsDetail).toBeCalledTimes(0)
      mockObjectsDetail.mockRestore()
    })
  })

  describe('options', () => {
    it('should return valid options', async () => {
      const field = TestModel.getField('user')
      const options = await field.options
      expect(options).toEqual({
        model: User,
        fieldName: 'name'
      })
    })

    it('should throw RequiredFieldOptionsException for missing options.model', async () => {
      const field = new ForeignKeyField({ options: {} })
      await expect(field.options).rejects.toBeInstanceOf(RequiredFieldOptionsException)
    })

    it('should throw RequiredFieldOptionsException for missing options.fieldName', async () => {
      const field = new ForeignKeyField({ options: { model: User } })
      await expect(field.options).rejects.toBeInstanceOf(RequiredFieldOptionsException)
    })

    it('should throw InvalidFieldOptionsException for missing options.fieldName', async () => {
      const field = new ForeignKeyField({ options: { model: {}, fieldName: 'name' } })
      await expect(field.options).rejects.toBeInstanceOf(InvalidFieldOptionsException)
    })
  })

  describe('prepareDisplayRender', () => {
    it('should return foreign field instance', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => user)

      const renderData = await field.prepareDisplayRender()

      expect(Object.keys(renderData)).toHaveLength(2)

      expect(renderData.field).toBe(user.getField('name'))
      expect(renderData.displayField).toBe(DisplayField)

      expect(mockObjectsDetail).toBeCalledTimes(1)
      mockObjectsDetail.mockRestore()
    })

    it('should return no field data', async () => {
      const model = new TestModel()
      const field = model.getField('user') as ForeignKeyField

      const renderData = await field.prepareDisplayRender()

      expect(renderData).toEqual({
        field: null,
        displayField: null
      })
    })
  })

  describe('displayRender', () => {
    it('should render correct value', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(
        async pk => pk === 1 ? userList[0] : userList[2])

      const wrapper = mount(BaseDisplayFieldRender, {
        propsData: { field: field }
      })

      await waitRender.DisplayField(wrapper)
      await waitRender.DisplayField(wrapper)
      expect(wrapper.html()).toBe('')

      expect(mockObjectsDetail).toBeCalledTimes(1)
      expect(mockObjectsDetail.mock.calls[0]).toEqual([1])

      await waitRender.DisplayField(wrapper)
      expect(wrapper.html()).toMatchSnapshot()

      model.val.user = 3
      expect(model.data).toEqual({ user: 3 })

      await waitRender.DisplayFieldUpdate(wrapper)
      await waitRender.DisplayFieldUpdate(wrapper)

      expect(mockObjectsDetail).toBeCalledTimes(2)
      expect(mockObjectsDetail.mock.calls[1]).toEqual([3])

      await waitRender.DisplayFieldUpdate(wrapper)

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsDetail.mockRestore()
    })
  })

  describe('mapInputSelectList', () => {
    it('should return mapped select list', async () => {
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)

      const options = await field.options as ForeignKeyFieldOptions
      const selectList = await field.mapInputSelectList(options)

      expect(selectList).toEqual([
        { value: '1', text: 'User 1' },
        { value: '2', text: 'User 2' },
        { value: '3', text: 'User 3' }
      ])

      expect(mockObjectsList).toBeCalledTimes(1)
      mockObjectsList.mockRestore()
    })

    it('should log warning if no primary key', async () => {
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => [
        ...userList,
        new User({ name: 'New User' })
      ])
      const mockConsoleWarn = jest.spyOn(console, 'warn').mockImplementation()

      const options = await field.options as ForeignKeyFieldOptions
      const selectList = await field.mapInputSelectList(options)

      expect(selectList).toEqual([
        { value: '1', text: 'User 1' },
        { value: '2', text: 'User 2' },
        { value: '3', text: 'User 3' },
        { value: 'null', text: 'New User' }
      ])

      expect(mockConsoleWarn).toBeCalledTimes(1)
      expect(mockConsoleWarn.mock.calls[0]).toEqual(['[vue-service-model] No primary key defined for model', 'User'])
      mockConsoleWarn.mockRestore()

      expect(mockObjectsList).toBeCalledTimes(1)
      mockObjectsList.mockRestore()
    })

    it('should log error when non string field is used for ForeignKeyField input', async () => {
      class TestModel extends BaseModel {
        static fieldsDef = {
          user: new ForeignKeyField({
            options: {
              model: User,
              fieldName: 'id'
            }
          })
        }
      }

      const model = new TestModel({ user: 1 })
      const field = model.getField('user') as ForeignKeyField

      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)
      const mockConsoleError = jest.spyOn(console, 'error').mockImplementation()

      const options = await field.options as ForeignKeyFieldOptions
      const selectList = await field.mapInputSelectList(options)

      expect(selectList).toEqual([
        { value: '1', text: 'unknown' },
        { value: '2', text: 'unknown' },
        { value: '3', text: 'unknown' }
      ])

      expect(mockConsoleError).toBeCalledTimes(3)
      expect(mockConsoleError.mock.calls[0]).toEqual([
        '[vue-service-model] Cannot use non string field for ForeignKeyField input. Used field name:',
        'id'
      ])
      mockConsoleError.mockRestore()

      expect(mockObjectsList).toBeCalledTimes(1)
      mockObjectsList.mockRestore()
    })
  })

  describe('inputRender', () => {
    it('should render correct input field', async () => {
      const model = new TestModel({ user: user.pk })
      const field = model.getField('user') as ForeignKeyField
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)

      const wrapper = mount(InputField, { propsData: { field } })
      await waitRender.InputField(wrapper)

      expect(mockObjectsList).toBeCalledTimes(1)

      await waitRender.InputField(wrapper)
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      const selectElement = wrapper.get('select')
      const options = selectElement.findAll('option')
      expect(options.at(0).attributes('selected')).toBe('selected')

      model.val.user = 3

      await waitRender.InputFieldUpdate(wrapper)

      expect(mockObjectsList).toBeCalledTimes(2)

      await waitRender.InputFieldUpdate(wrapper)
      await wrapper.vm.$nextTick()
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsList.mockRestore()
    })

    it('should render correct input field without value', async () => {
      const model = new TestModel({})
      const field = model.getField('user') as ForeignKeyField
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)

      const wrapper = mount(InputField, { propsData: { field } })
      await waitRender.InputField(wrapper)

      expect(mockObjectsList).toBeCalledTimes(1)

      await waitRender.InputField(wrapper)
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsList.mockRestore()
    })

    it('should set value correct on change', async () => {
      const model = new TestModel({ user: user.pk })
      const field = model.getField('user') as ForeignKeyField

      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => userList[2])

      const wrapper = mount(InputField, { propsData: { field } })
      await waitRender.InputField(wrapper)

      expect(mockObjectsList).toBeCalledTimes(1)

      await waitRender.InputField(wrapper)
      await wrapper.vm.$nextTick()

      const selectElement = wrapper.get('select')
      const options = selectElement.findAll('option')
      expect(options.at(0).attributes('selected')).toBe('selected')

      const newOption = options.at(2).element as HTMLOptionElement

      newOption.selected = true
      selectElement.trigger('input')

      await waitRender.InputFieldUpdate(wrapper)
      await waitRender.InputFieldUpdate(wrapper)
      await waitRender.InputFieldUpdate(wrapper)
      await waitRender.InputFieldUpdate(wrapper)
      await wrapper.vm.$nextTick()
      await wrapper.vm.$nextTick()
      await wrapper.vm.$nextTick()
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()
      expect(options.at(2).attributes('selected')).toBe('selected')

      expect(mockObjectsDetail).toBeCalledTimes(0)

      const selectedUser = await model.val.user
      expect(selectedUser).toBe(userList[2])
      expect(model.data).toEqual({ user: '3' })

      expect(mockObjectsDetail).toBeCalledTimes(1)
      expect(mockObjectsDetail.mock.calls[0]).toEqual(['3'])

      expect(mockObjectsList).toBeCalledTimes(2)

      mockObjectsDetail.mockRestore()
      mockObjectsList.mockRestore()
    })

    it('should render disabled input field', async () => {
      const model = new TestModel({ user: user.pk })
      const field = model.getField('user') as ForeignKeyField
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)

      const wrapper = mount(InputField, { propsData: { field, disabled: true } })
      await waitRender.InputField(wrapper)

      expect(mockObjectsList).toBeCalledTimes(1)

      await waitRender.InputField(wrapper)
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsList.mockRestore()
    })

    it('should render readonly input field', async () => {
      const model = new TestModel({ user: user.pk })
      const field = model.getField('user') as ForeignKeyField
      const mockObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(async () => userList)

      const wrapper = mount(InputField, { propsData: { field, readonly: true } })
      await waitRender.InputField(wrapper)

      expect(mockObjectsList).toBeCalledTimes(1)

      await waitRender.InputField(wrapper)
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsList.mockRestore()
    })
  })
})
