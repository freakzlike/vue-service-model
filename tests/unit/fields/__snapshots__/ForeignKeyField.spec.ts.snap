// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields/ForeignKeyField displayRender should render correct value 1`] = `<div>User 1</div>`;

exports[`fields/ForeignKeyField displayRender should render correct value 2`] = `<div>User 3</div>`;

exports[`fields/ForeignKeyField inputRender should render correct input field 1`] = `
<select>
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;

exports[`fields/ForeignKeyField inputRender should render correct input field 2`] = `
<select>
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;

exports[`fields/ForeignKeyField inputRender should render correct input field without value 1`] = `
<select>
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;

exports[`fields/ForeignKeyField inputRender should render disabled input field 1`] = `
<select disabled="">
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;

exports[`fields/ForeignKeyField inputRender should render readonly input field 1`] = `
<select readonly="">
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;

exports[`fields/ForeignKeyField inputRender should set value correct on change 1`] = `
<select>
  <option value="1">User 1</option>
  <option value="2">User 2</option>
  <option value="3">User 3</option>
</select>
`;
